AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestMatrixStore.kt

```diff
@@ -105,21 +105,89 @@ internal class TestMatrixStore(
         }
-        val newTestMatrix = firebaseTestLabApi.createTestMatrix(
-            projectId = firebaseProjectId,
-            requestId = UUID.randomUUID().toString(),
-            testMatrix = createNewTestMatrix(
-                testRunKey = testRunId,
-                environmentMatrix = environmentMatrix,
-                clientInfo = clientInfo,
-                sharding = sharding,
-                deviceSetup = deviceSetup,
-                appApk = appApk,
-                testApk = testApk,
-                pullScreenshots = pullScreenshots,
-                testTargets = testTargets,
-                flakyTestAttempts = flakyTestAttempts
-            )
+
+        val newTestMatrix = createNewTestMatrix(
+            testRunKey = testRunId,
+            environmentMatrix = environmentMatrix,
+            clientInfo = clientInfo,
+            sharding = sharding,
+            deviceSetup = deviceSetup,
+            appApk = appApk,
+            testApk = testApk,
+            pullScreenshots = pullScreenshots,
+            testTargets = testTargets,
+            flakyTestAttempts = flakyTestAttempts
+        )
+        logger.info {
+            "created test matrix: $newTestMatrix"
+        }
+        // save it to cache, we don't worry about races here much such that if another instance happens to be creating
+        // the exact same test, one of them will win but that is OK since they'll each use their own test matrices and
+        // followup calls will re-use the winner of this race.
+        val testRun = TestRun(
+            id = testRunId,
+            testMatrixId = checkNotNull(newTestMatrix.testMatrixId) {
+                "newly created test matrix should not have null id $newTestMatrix"
+            }
+        )
+        datastoreApi.put(testRun.toEntity())
+        logger.info {
+            "saved test matrix info: $testRun"
+        }
+        return newTestMatrix
+    }
+
+    /**
+     * Creates a TestMatrix for the given configuration or returns an existing one if we've run the same tests
+     * specified in [testTargets] with the same environment configuration.
+     */
+    suspend fun getOrCreateTestMatrix(
+        testMatrix: TestMatrix,
+        cachedTestMatrixFilter: CachedTestMatrixFilter = { true },
+        testTargets: List<String>? = null,
+        flakyTestAttempts: Int = 2
+    ): TestMatrix {
+        checkNotNull(testMatrix.testMatrixId) {
+            "Test matrix id for the base test matrix should not be null"
+        }
+        val testRunId = TestRun.createId(
+            datastoreApi = datastoreApi,
+            environment = testMatrix.environmentMatrix,
+            clientInfo = testMatrix.clientInfo,
+            sharding = testMatrix.testSpecification.androidInstrumentationTest?.shardingOption,
+            testSetup = testMatrix.testSpecification.testSetup,
+            testTargets = testTargets,
+            testMatrixId = testMatrix.testMatrixId
```

Comment:
lets name this: "baseTestMatrixId" or something like that to avoid confusion with TestRunId -> TestMatrixId association.

Reply:
done